File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int lo=0, hi=0, n=(int)s.length(), ans=0;
        vector<int> vec(256,-1);
        while(hi<n){
            if(vec[s[hi]]==-1){
                vec[s[hi]]=hi;
            }
            else{
                int idx=vec[s[hi]];
                while(lo<=idx){
                    vec[s[lo++]]=-1;
                }
                vec[s[hi]]=hi;
            }
            ans=max(ans,hi-lo+1);
            ++hi;
        } 
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> cache = new HashSet<>();
        int left=0, ans=0;
        for(int right=0;right<s.length();right++){
            char ch= s.charAt(right);
            while(cache.contains(ch)){
                cache.remove(s.charAt(left++));
            }
            cache.add(ch);
            ans= Math.max(ans,right-left+1);
        }
        return ans;
    }
}